# Day 67 – Introduction to MLOps and MLflow Basics

## Introduction

In this notebook, I explored the **foundations of MLOps (Machine Learning Operations)** and how it helps bridge the gap between machine learning development and deployment.  
The main focus of this practical is on using **MLflow**, an open-source MLOps tool, to **track experiments, log metrics, and manage models** efficiently.  

I implemented a small experiment using a **Logistic Regression** model on a synthetic dataset and integrated MLflow for experiment tracking.  

This notebook marks the beginning of my MLOps journey — understanding how to make ML workflows **reproducible, traceable, and production-ready**.

---

## Introduction to MLOps

## What is MLOps?

**MLOps (Machine Learning Operations)** is the process of managing the entire machine learning lifecycle — from model building to deployment and monitoring — in a systematic, automated, and scalable way.

Just like **DevOps** for software development, **MLOps** brings collaboration between:
- Data Scientists (who build models)  
- ML Engineers (who deploy and manage them)  

The main goal of MLOps is to make machine learning models **reproducible, reliable, and deployable at scale**.

---

## Why MLOps is Important

Without MLOps, ML models often:
- Stay stuck in notebooks and never reach production.
- Are difficult to monitor or update after deployment.
- Fail silently when data or trends change.

MLOps solves these problems by:
- Making deployment faster and automated  
- Ensuring collaboration between **data scientists** and **DevOps teams**  
- Improving **model reproducibility and scalability**  
- Enabling **continuous retraining** as data evolves  

In short — **MLOps = Machine Learning + DevOps**

---

## Key Components of MLOps

1. **Version Control** – Track code, data, and model versions (e.g., Git, DVC).  
2. **Continuous Integration (CI)** – Automatically test and validate model code.  
3. **Model Deployment** – Serve trained models as APIs or microservices.  
4. **Monitoring** – Track accuracy, latency, and data drift after deployment.  
5. **Retraining** – Automatically update models when performance drops.  

---


## MLOps Lifecycle

The end-to-end workflow of MLOps involves the following stages:

1. **Data Collection & Preparation** – Gathering, cleaning, and transforming data.  
2. **Model Development** – Training, tuning, and validating models.  
3. **Experiment Tracking** – Logging parameters, metrics, and model artifacts.  
4. **Model Deployment** – Serving models through APIs or cloud services.  
5. **Monitoring & Retraining** – Tracking model drift and retraining periodically.

Each stage is connected, making ML systems automated and production-ready.


---

## Common Tools Used in MLOps

| Tool | Purpose |
|------|----------|
| **Git/DVC** | Version control for code and data |
| **MLflow** | Track experiments, models, and parameters |
| **Docker** | Containerize models for deployment |
| **Kubeflow** | Manage ML workflows on Kubernetes |
| **TensorFlow Extended (TFX)** | Build production-ready ML pipelines |
| **Cloud Platforms (AWS, Azure, GCP)** | Scalable hosting, storage, and deployment |


## Introduction to MLflow

## What is MLflow?

**MLflow** is an open-source MLOps tool that helps manage the ML lifecycle.  
It provides simple APIs to:

-  **Track Experiments:** Log parameters, metrics, and artifacts.  
-  **Package Models:** Save trained models in a standard format.  
-  **Deploy Models:** Easily deploy models to production or cloud.  
-  **Register Models:** Keep multiple model versions and promote the best one.

MLflow is lightweight, beginner-friendly, and works with any ML library (Scikit-learn, XGBoost, PyTorch, TensorFlow, etc.).

---


## Running MLflow UI for Experiment Tracking

Before accessing the MLflow interface in the browser, we need to **start the MLflow tracking server**.

---

## Run MLflow UI from Terminal

```bash
mlflow ui
````

### What happens:

* This command launches a **local MLflow tracking server**.
* By default, it runs at: [http://127.0.0.1:5000](http://127.0.0.1:5000)
* The notebook logs all your **experiments, parameters, metrics, and models** into this server for easy viewing.

> **Note:**
> Keep this terminal window open while using MLflow.
> Closing it will stop the tracking server and the UI will no longer be accessible.

---

## Why We Need MLflow UI

The MLflow UI acts as your **central dashboard** for managing experiments. It allows you to:

* Organize experiments under named **projects (Experiments)**
* Log and compare **runs** (each training session = one run)
* Evaluate models based on **metrics** (accuracy, RMSE, F1-score, etc.)
* View and download **artifacts** such as trained models, plots, and requirements files
* Track **parameters and hyperparameters** across multiple runs

In short — **MLflow UI is your control center** for tracking, comparing, and improving your models.

---

## How the Links Work

After running your MLflow experiment, two important links appear in the notebook output:

> **Link 1 – MLflow Home:**
> Opens the dashboard listing all experiments.
> Example: [http://127.0.0.1:5000](http://127.0.0.1:5000)

> **Link 2 – Direct Run URL:**
> Takes you directly to the latest run’s details (metrics, parameters, artifacts).

Both links open in your browser, allowing you to visualize and analyze your ML experiments easily.


---

## Import Libraries

We start by importing all the necessary Python libraries.

- **numpy** → for numerical operations and array handling  
- **make_classification** → creates a synthetic dataset for ML testing  
- **train_test_split** → splits data into training and testing sets  
- **LogisticRegression** → simple classification algorithm from sklearn  
- **classification_report** → evaluates model performance using metrics  
- **warnings** → suppresses unwanted warnings for cleaner output  
- **mlflow** → used to log and track experiments, metrics, and models  


In [1]:
import numpy as np
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import classification_report
import warnings
warnings.filterwarnings('ignore')

## Dataset Creation

We will generate a **synthetic imbalanced dataset** using sklearn’s `make_classification`.  
This simulates a real-world case (like fraud detection) where one class appears much less frequently.

- `n_samples = 1000` → total rows  
- `n_features = 10` → number of columns  
- `n_informative = 2` → actual useful features  
- `n_redundant = 8` → noisy or irrelevant features  
- `weights = [0.9, 0.1]` → imbalance ratio (90% class 0, 10% class 1)  
- `random_state = 42` → ensures reproducible results  


In [2]:
# Step 1: Create an imbalanced binary classification dataset
X, y = make_classification(n_samples=1000, n_features=10, n_informative=2, n_redundant=8, 
                           weights=[0.9, 0.1], flip_y=0, random_state=42)

np.unique(y, return_counts=True)

(array([0, 1]), array([900, 100], dtype=int64))

## Train-Test Split

We split the dataset into training and testing subsets.

- **Training set** → used to fit the model  
- **Testing set** → used to evaluate on unseen data  
- `test_size = 0.3` → 30 % for testing, 70 % for training  
- `stratify = y` → keeps the same class ratio in both splits  


In [3]:
# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, stratify=y, random_state=42)

## Define and Train the Model

We’ll use **Logistic Regression**, a basic yet powerful algorithm for binary classification.

**Model parameters:**
- `solver = "lbfgs"` → optimization algorithm  
- `max_iter = 1000` → ensures convergence  
- `multi_class = "auto"` → automatically handles binary/multi-class data  
- `random_state = 42` → reproducibility  

The model is then trained (fitted) on the training data.  


In [8]:
# Define the model hyperparameters
params = {
    "solver": "lbfgs",
    "max_iter": 1000,
    "multi_class": "auto",
    "random_state": 42,
}

# Train the model
lr = LogisticRegression(**params)
lr.fit(X_train, y_train)

LogisticRegression(max_iter=1000, random_state=42)

## Predictions and Evaluation

After training, we predict on the test data and evaluate performance.

`classification_report` provides:  
- **Precision** → percentage of correct positive predictions  
- **Recall** → ability to detect actual positives  
- **F1-score** → harmonic mean of precision & recall  
- **Accuracy** → overall proportion of correct predictions  


In [9]:
# Predict on the test set
y_pred = lr.predict(X_test)

report = classification_report(y_test, y_pred)
print(report)

              precision    recall  f1-score   support

           0       0.95      0.97      0.96       270
           1       0.62      0.50      0.56        30

    accuracy                           0.92       300
   macro avg       0.79      0.73      0.76       300
weighted avg       0.91      0.92      0.92       300



In [5]:
report_dict = classification_report(y_test, y_pred, output_dict=True)
report_dict

{'0': {'precision': 0.9456521739130435,
  'recall': 0.9666666666666667,
  'f1-score': 0.956043956043956,
  'support': 270.0},
 '1': {'precision': 0.625,
  'recall': 0.5,
  'f1-score': 0.5555555555555556,
  'support': 30.0},
 'accuracy': 0.92,
 'macro avg': {'precision': 0.7853260869565217,
  'recall': 0.7333333333333334,
  'f1-score': 0.7557997557997558,
  'support': 300.0},
 'weighted avg': {'precision': 0.9135869565217392,
  'recall': 0.92,
  'f1-score': 0.915995115995116,
  'support': 300.0}}

## Experiment Tracking with MLflow

Next, we integrate **MLflow** to log and track our experiments.

- `mlflow.set_experiment("1st Experiment")` → creates/sets the experiment name  
- `mlflow.set_tracking_uri("http://127.0.0.1:5000/")` → connects to local tracking server  
- Inside the `mlflow.start_run()` block, we log:  
  - **Parameters** → via `mlflow.log_params()`  
  - **Metrics** → via `mlflow.log_metrics()`  
  - **Model artifact** → via `mlflow.sklearn.log_model()`  

This makes our training runs reproducible and comparable through the MLflow UI.  


In [6]:
import mlflow

In [7]:
mlflow.set_experiment("1st Experiment")
mlflow.set_tracking_uri(uri="http://127.0.0.1:5000/")
#mlflow.set_tracking_uri("http://localhost:5000")

with mlflow.start_run():
    mlflow.log_params(params)
    mlflow.log_metrics({
        'accuracy': report_dict['accuracy'],
        'recall_class_0': report_dict['0']['recall'],
        'recall_class_1': report_dict['1']['recall'],
        'f1_score_macro': report_dict['macro avg']['f1-score']
    })
    mlflow.sklearn.log_model(lr, "Logistic Regression")  

2025/11/11 20:28:50 INFO mlflow.tracking.fluent: Experiment with name '1st Experiment' does not exist. Creating a new experiment.
2025/11/11 20:28:50 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2025/11/11 20:29:08 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


🏃 View run sedate-squirrel-428 at: http://127.0.0.1:5000/#/experiments/838503554236628539/runs/6a4e577acc054661b623682852c93b78
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/838503554236628539


# MLflow UI – Experiment Tracking Overview

This section visually demonstrates how the **MLflow Tracking UI** organizes experiments, runs, and metrics.  
Each image below corresponds to a specific part of the MLflow interface.

---

## 1. Experiments Dashboard

![MLflow Experiments Dashboard](screenshot1.png)

**Explanation:**  
This is the **main Experiments Dashboard** of MLflow.  
It lists all the experiments created while running your machine learning projects.  
Here you can:  
- View all experiments (like *1st Experiment* and *Default*)  
- Check creation time, modification date, and tags  
- Create new experiments manually using the **Create** button  
- Filter or search experiments easily by name  

This serves as the **starting point** for navigating into specific experiments and their runs.

---

## 2. Runs List under an Experiment

![MLflow Runs List](screenshot2.png)

**Explanation:**  
After selecting **1st Experiment**, you see all the **runs** performed under it.  
Each row represents a single training run executed from the notebook.  
It shows:  
- **Run name** (e.g., *sedate-squirrel-428*)  
- **Creation time**, **Duration**, and **Source file**  
- Linked **Model artifact** (e.g., *Logistic Regression*)  

You can also sort, filter, or compare runs to analyze model performance differences.

---

## 3. Run Overview – Metrics and Parameters Summary

![MLflow Run Overview](screenshot3.png)

**Explanation:**  
This page summarizes everything about one specific run.  
It displays:  
- **Metrics:** accuracy, F1-score, recall for both classes  
- **Parameters:** model hyperparameters like solver, max_iter, and random_state  
- **Run metadata:** Run ID, Experiment ID, Duration, and Status (*Finished*)  
- **Logged model:** Shows the associated ML model (here, *Logistic Regression*)  

This view helps verify whether your experiment executed correctly and what values were logged.

---

## 4. Run Details – Parameters and Logged Model

![MLflow Parameters and Logged Model](screenshot4.png)

**Explanation:**  
This view provides detailed **parameter values** and information about the **saved model**.  
It includes:  
- List of parameters used (max_iter, multi_class, random_state, solver)  
- Section for logged models with type, step, and creation time  
- “Ready” status confirms the model was successfully logged  

This screen confirms reproducibility — anyone can rerun the model with the same settings.

---

## 5. Model Metrics Visualization

![MLflow Model Metrics](screenshot5.png)

**Explanation:**  
This is the **Model Metrics tab**, showing visualized metric charts for each tracked metric.  
In this run, the metrics include:  
- **Accuracy:** 0.92  
- **F1 Score (Macro):** 0.76  
- **Recall (Class 0):** 0.97  

Each blue bar represents a metric value for the selected run.  
This visualization helps quickly compare performance across different experiments or runs.

---

# Summary

These screenshots collectively demonstrate how MLflow tracks:
- Experiment-level organization  
- Run-level metrics and parameters  
- Model artifact logging  
- Visual comparison of performance  

MLflow’s intuitive interface makes it easy to **analyze, reproduce, and compare** machine learning experiments efficiently.


---

## Conclusion

Through this notebook, I successfully implemented my **first MLOps experiment** using **MLflow**.  
I learned how to:
- Create and train a machine learning model.  
- Log **parameters, metrics, and model artifacts** automatically.  
- Use the **MLflow UI** to explore experiment runs visually.  

This exercise gave me a clear idea of how data scientists and ML engineers collaborate through tools like MLflow to streamline the ML lifecycle.  
The entire process—from training to tracking—was automated and well-documented, showing how MLOps improves workflow transparency.


## Key Learnings

In this Day 67 notebook, I learned and implemented:  
- The **concept and importance of MLOps** in modern machine learning workflows.  
- Basics of **MLflow**, including its components and purpose.  
- How to run MLflow UI and log experiments programmatically.  
- Visualization and comparison of runs through **MLflow Dashboard**.  
- Practical experience in **experiment tracking and model versioning**.  

This marks the foundation for future MLOps tasks such as **model registry, deployment, and monitoring**, which I’ll continue exploring in upcoming notebooks.

---